src/bundle_info.c

```diff
@@ -0,0 +1,553 @@
+		}
+
+		if (manifest->optional) {
+			get_indirect_includes(mom, manifest->optional, direct_includes, indirect_includes, true);
+		}
+
+		free_manifest(manifest);
+	}
+
+	return SWUPD_OK;
+}
+
+static enum swupd_code get_included_bundles(struct bundle_info *bundle_info, struct manifest *mom)
+{
+	struct manifest *bundle_manifest = bundle_info->manifest;
+	struct list *direct_includes = NULL;
+	struct list *indirect_includes = NULL;
+	enum swupd_code ret;
+
+	/* get a list of all includes (direct and indirect includes) */
+	direct_includes = list_clone_and_concat(bundle_manifest->includes, bundle_manifest->optional);
+	direct_includes = list_sort(direct_includes, list_strcmp);
+	ret = get_indirect_includes(mom, direct_includes, direct_includes, &indirect_includes, false);
+	if (ret) {
+		list_free_list(direct_includes);
+		list_free_list(indirect_includes);
+		return ret;
+	}
+
+	/* some bundles could have included bundles that had already
+	 * been included by another bundle, so remove duplication */
+	indirect_includes = list_string_deduplicate(indirect_includes);
+
+	bundle_info->direct_includes = direct_includes;
+	bundle_info->indirect_includes = indirect_includes;
+
+	return ret;
+}
+
+static enum swupd_code get_bundle_size(struct bundle_info *bundle_info)
+{
+	struct file *file = NULL;
+	struct list *iter;
+	struct stat st;
+	enum swupd_code ret;
+	long size_installed = 0;
+
+	/* there are two different sizes we are interested here:
+	 *  - the total size of the bundle including all its dependencies,
+	 *  - the size increase if the bundle gets installed (if already installed the increase is 0):
+	 *    increase = <bundle size> + <included bundles size> - <size of files from bundle and dependencies already installed> */
+	bundle_info->size_total = get_manifest_list_contentsize(bundle_info->includes_manifests);
+
+	/* subtract the size of the files already installed from the total size to
+	 * figure out how much space is required for the bundle if not installed */
+	iter = list_head(bundle_info->files);
+	while (iter) {
+		file = iter->data;
+		iter = iter->next;
+
+		if (lstat(file->filename, &st) == 0) {
+			if (S_ISREG(st.st_mode)) {
+				size_installed += st.st_size;
+			}
+		}
+	}
+	if (bundle_info->size_total - size_installed < 0) {
+		ret = SWUPD_UNEXPECTED_CONDITION;
+	}
+	bundle_info->size_required = bundle_info->size_total - size_installed;
+
+	return ret;
+}
+
+enum swupd_code bundle_info(char *bundle)
+{
+	enum swupd_code ret = SWUPD_OK;
+	int requested_version, latest_version;
+	struct manifest *mom = NULL;
+	struct manifest *latest_mom = NULL;
+	struct manifest *latest_manifest = NULL;
+	struct file *file = NULL;
+	struct list *subs;
+	struct list *iter;
+	struct bundle_info bundle_info = { NULL, NULL, NULL, NULL, NULL, false, false, 0, 0, 0 };
+	bool mix_exists;
+	bool system_up_to_date;
+
+	/* get the very latest version available */
+	latest_version = version_get_absolute_latest();
+	if (latest_version < 0) {
```

Comment:
Depending on whats get merged first,this or #1076  . Worthwhile taking a note to check for SWUPD_SIGNATURE_VERIFICATION_FAILED as well

Reply:
yeah, whatever PR merges first will force a rebase in the other PR cause the tests will fail if not rebased